File: Android/app/src/main/java/com/hhp227/datemate/data/repository/StorageRepository.kt
package com.hhp227.datemate.data.repository

import android.net.Uri
import android.util.Log
import com.hhp227.datemate.common.Resource
import com.hhp227.datemate.data.datasource.StorageRemoteDataSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.sync.Semaphore

class StorageRepository private constructor(
    private val storageRemoteDataSource: StorageRemoteDataSource
) {
    private suspend fun uploadProfileImage(uri: Uri, userId: String, index: Int): String {
        val path = "users/$userId/gallery_${index}_${System.currentTimeMillis()}.jpg"
        return storageRemoteDataSource.uploadFile(uri, path)
    }

    fun uploadAllImages(imageUris: List<Uri>, userId: String, concurrency: Int = 3): Flow<Resource<List<String>>> = flow {
        try {
            val semaphore = Semaphore(concurrency)
            val urls = coroutineScope {
                imageUris.mapIndexed { index, uri ->
                    async(Dispatchers.IO) {
                        semaphore.acquire()
                        try {
                            uploadProfileImage(uri, userId, index)
                        } finally {
                            semaphore.release()
                        }
                    }
                }.awaitAll()
            }

            emit(Resource.Success(urls))
        } catch (e: Exception) {
            emit(Resource.Error<List<String>>(e.message ?: "Upload failed"))
        }
    }
        .onStart { emit(Resource.Loading()) }

    companion object {
        @Volatile private var instance: StorageRepository? = null

        fun getInstance(remote: StorageRemoteDataSource): StorageRepository =
            instance ?: synchronized(this) {
                instance ?: StorageRepository(remote).also { instance = it }
            }
    }
}

File: iOS/DateMate/Data/Repository/StorageRepository.swift
//
//  StorageRepository.swift
//  DateMate
//
//  Created by 홍희표 on 2025/11/26.
//

import Foundation
import Combine

class StorageRepository {
    private let storageRemoteDataSource: StorageRemoteDataSource
    
    private let defaultConcurrency: Int
    
    func uploadProfileImagePublisher(fileUrl: URL, userId: String, index: Int) -> AnyPublisher<Resource<String>, Never> {
        let path = "users/\(userId)/gallery_\(index)_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
        return storageRemoteDataSource.uploadFile(fileUrl: fileUrl, path: path).asResource()
    }
    
    func uploadAllImages(
        imageUrls: [URL],
        userId: String,
        concurrency: Int? = nil,
        retryCount: Int = 0
    ) -> AnyPublisher<[String], Error> {
        let concurrency = concurrency ?? defaultConcurrency
        let publishers = imageUrls.enumerated().map { (index, url) -> AnyPublisher<String, Error> in
            let path = "users/\(userId)/gallery_\(index)_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
            return storageRemoteDataSource.uploadFile(fileUrl: url, path: path)
                .retry(retryCount)
                .eraseToAnyPublisher()
        }
        return Publishers.MergeMany(publishers)
            .collect()
            .eraseToAnyPublisher()
    }
    
    private init(_ storageRemoteDataSource: StorageRemoteDataSource, defaultConcurrency: Int = 4) {
        self.storageRemoteDataSource = storageRemoteDataSource
        self.defaultConcurrency = defaultConcurrency
    }
    
    private static var instance: StorageRepository? = nil
    
    static func getInstance(storageRemoteDataSource: StorageRemoteDataSource) -> StorageRepository {
        if let instance = self.instance {
            return instance
        } else {
            let storageRepository = StorageRepository(storageRemoteDataSource)
            self.instance = storageRepository
            return storageRepository
        }
    }
}
